### PR TITLE
HPCC-15432 Avoid conditionals child activity early initialization

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith jake.smith@lexisnexis.com

### CODE REVIEW

**Comment by jakesmith** [comment]
@ghalliday - please review

File: thorlcr/activities/diskread/thdiskreadslave.cpp [hunk 1]
```diff
@@ -412,6 +412,7 @@ class CDiskReadSlaveActivity : public CDiskReadSlaveActivityRecord
         else
             limit = (rowcount_t)helper->getRowLimit();
```
Comment (ghalliday) [inline thorlcr/activities/diskread/thdiskreadslave.cpp]: Not part of this pull request - but these can't be called in the constructor, they need to be in start()

File: thorlcr/activities/diskread/thdiskreadslave.cpp [hunk 2]
```diff
@@ -601,6 +599,7 @@ class CDiskNormalizeSlave : public CDiskReadSlaveActivityRecord
...
         stopAfter = (rowcount_t)helper->getChooseNLimit();
```
Comment (ghalliday) [inline thorlcr/activities/diskread/thdiskreadslave.cpp]: same problem here, accessing helper functions too early

File: thorlcr/activities/diskread/thdiskreadslave.cpp [hunk 3]
```diff
@@ -846,6 +844,7 @@ class CDiskCountSlave : public CDiskReadSlaveActivityRecord
         mpTag = TAG_NULL;
         stopAfter = (rowcount_t)helper->getChooseNLimit();
```
Comment (ghalliday) [inline thorlcr/activities/diskread/thdiskreadslave.cpp]: and here

File: thorlcr/activities/project/thprojectslave.cpp [hunk 4]
```diff
@@ -255,10 +251,10 @@ class CPrefetchProjectSlaveActivity : public CSlaveActivity
         preload = helper->getLookahead();
```
Comment (ghalliday) [inline thorlcr/activities/project/thprojectslave.cpp]: Theoretically this could also be dynamic (e.g., based on a stored variable) => move to start()

File: thorlcr/activities/xmlread/thxmlreadslave.cpp [hunk 5]
```diff
@@ -204,6 +204,7 @@ class CXmlReadSlaveActivity : public CDiskReadSlaveActivityBase
             limit = RCMAX;
         else
             limit = (rowcount_t)helper->getRowLimit();
```
Comment (ghalliday) [inline thorlcr/activities/xmlread/thxmlreadslave.cpp]: Another helper access that should be in start() (and a couple of lines before)

**Comment by ghalliday** [comment]
@jakesmith some incidental issues picked up when reviewing.

I was slightly concerned that this would cause more data to be sent, but as far as I can understand it shouldn't make any difference.

**Comment by jakesmith** [comment]
I have addressed the incidental issues (helper calls being called too early) in new commit.

>I was slightly concerned that this would cause more data to be sent, but as far as I can understand it shouldn't make any difference.

No it should not make any difference. It will initialize and serialize new arms of the graph if they are being re-executed and have not been initialized/serialized before.

@ghalliday - please check new commit
